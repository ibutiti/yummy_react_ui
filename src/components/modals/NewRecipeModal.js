import React, { Component } from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Modal, Icon, Card } from 'semantic-ui-react';

import RecipeForm from '../forms/RecipeForm';
import { createRecipe } from '../../actions/recipes';
import { fetchCategories } from '../../actions/categories';

class NewRecipeModal extends Component {
  state = {
    modalOpen: false
  };

  componentDidMount() {
    this.props.fetchCategories(1, 1000, this.props.apiKey);
  }

  modalClick = () => {
    this.setState({ modalOpen: !this.state.modalOpen });
  };

  submit = data =>
    this.props.createRecipe(data, this.props.apiKey).then(() => {
      this.setState({ modalOpen: false });
    });

  render() {
    return (
      <div>
        <Modal
          basic
          size="small"
          open={this.state.modalOpen}
          style={{ marginTop: '180px', marginLeft: '350px' }}
          trigger={
            <span onClick={this.modalClick}>
              <Icon name="add circle" className="category-modal-icon" />
              Add a Recipe
            </span>
          }>
          <Modal.Content>
            <Modal.Description>
              <Card color="violet" fluid>
                <Card.Content>
                  <Card.Header>
                    {_.isEmpty(this.props.categories) ? (
                      <span onClick={this.modalClick}>
                        Create a Recipe Category First
                      </span>
                    ) : (
                      'Add a Recipe'
                    )}
                  </Card.Header>
                  {_.isEmpty(this.props.categories) ? (
                    ''
                  ) : (
                    <RecipeForm
                      categories={this.props.categories}
                      submit={this.submit}
                      cancel={this.modalClick}
                    />
                  )}
                </Card.Content>
              </Card>
            </Modal.Description>
          </Modal.Content>
        </Modal>
      </div>
    );
  }
}
NewRecipeModal.propTypes = {
  createRecipe: PropTypes.func.isRequired,
  fetchCategories: PropTypes.func.isRequired,
  apiKey: PropTypes.string.isRequired
};

function mapStateToProps(state) {
  return {
    apiKey: state.user.apikey,
    categories: state.categories.data
  };
}

export default connect(mapStateToProps, { createRecipe, fetchCategories })(
  NewRecipeModal
);
