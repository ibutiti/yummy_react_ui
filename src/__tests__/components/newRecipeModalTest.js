import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ReactDom from 'react-dom';
import configureMockStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import NewRecipeModal from '../../components/modals/NewRecipeModal';

Enzyme.configure({ adapter: new Adapter() });

describe('<NewRecipeModal>', () => {
  const store = configureMockStore([thunk])({
    recipes: { links: {}, data: {} },
    categories: { links: {}, data: {} },
    user: { apikey: 'abcd' }
  });
  const props = {
    modalClick: jest.fn(),
    createRecipe: jest.fn(() => Promise.resolve('recipes')),
    fetchCategories: jest.fn(() => Promise.resolve('categories')),
    submit: jest.fn(),
    setState: jest.fn(),
    modalOpen: false
  };

  // test it renders without crashing
  it('should render without crashing', () => {
    const { enzymeWrapper } = mount(
      <Provider store={store}>
        <NewRecipeModal {...props} />
      </Provider>
    );
  });

  it('renders fully', () => {
    const tree = renderer
      .create(
        mount(
          <Provider store={store}>
            <NewRecipeModal {...props} />
          </Provider>
        )
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
