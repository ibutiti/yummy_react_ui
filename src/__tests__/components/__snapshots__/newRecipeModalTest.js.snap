// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewRecipeModal> renders fully 1`] = `
<div>
  <span
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <i
      aria-hidden="true"
      className="add circle icon category-modal-icon"
    />
    Add a Recipe
  </span>
</div>
`;
